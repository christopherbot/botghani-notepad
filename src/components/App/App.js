import React, { PureComponent } from 'react'
import { View } from 'react-native'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { setActiveList } from '../../state/actions'

import Nav from '../Nav/Nav'
import ListHeader from '../ListHeader/ListHeader'
import List from '../List/List'
import NavButton from '../Buttons/NavButton/NavButton'

import style from './App.style'
import ustyle from '../../utils/style'

class App extends PureComponent {
  static propTypes = {
    isNavOpen: PropTypes.bool.isRequired,
  }

  static navigationOptions = () => {
    return {
      headerTitle : <ListHeader />,
      headerLeft : <NavButton />,
    }
  }

  componentDidUpdate(prevProps) {
    if (this.props.lists.length < prevProps.lists.length) {
      this.props.setActiveList(null)
    } else if (this.props.lists.length > prevProps.lists.length) {
      this.props.setActiveList(this.props.lists[this.props.lists.length - 1].id)
    }
  }

  render() {
    return (
      <View style={ustyle.f1}>
        <View style={style.navWrapper}>
          { this.props.isNavOpen && <Nav /> }
        </View>
        { this.props.list && <List list={this.props.list} /> }
      </View>
    )
  }
}

const mapStateToProps = ({ lists, globalUi }) => ({
  isNavOpen: globalUi.isNavOpen,
  lists,
  list: lists.find(list => list.id === globalUi.activeListId),
})

const mapDispatchToProps = {
  setActiveList,
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
