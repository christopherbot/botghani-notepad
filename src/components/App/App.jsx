import React, { PureComponent } from 'react'
import { View, Text } from 'react-native'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import {
  setActiveList,
  deleteList,
  setIsModalOpen,
} from 'state/actions'
import { listPropType } from 'components/propTypeDefinitions'

import List from 'components/List/List'
import Modal from 'components/Modal/Modal'
import HappyPaper from 'assets/happyPaper.svgx'
import SurprisedPaper from 'assets/surprisedPaper.svgx'

import gStyle from 'styles/globalStyle'
import style from './App.style'

export class App extends PureComponent {
  static propTypes = {
    lists: PropTypes.arrayOf(listPropType).isRequired,
    list: listPropType,
    favouriteListId: PropTypes.string,
    setActiveList: PropTypes.func.isRequired,
    deleteList: PropTypes.func.isRequired,
    closeModal: PropTypes.func.isRequired,
    isModalOpen: PropTypes.bool.isRequired,
    navigation: PropTypes.shape({
      openDrawer: PropTypes.func.isRequired,
      closeDrawer: PropTypes.func.isRequired,
    }).isRequired,
  }

  componentDidUpdate(prevProps) {
    if (this.props.lists.length < prevProps.lists.length) {
      this.props.setActiveList(null)
    } else if (this.props.lists.length > prevProps.lists.length) {
      this.props.setActiveList(this.props.lists[this.props.lists.length - 1].id)
    }
  }

  componentDidMount() {
    if (this.props.favouriteListId) {
      this.props.setActiveList(this.props.favouriteListId)
    }
  }

  deleteList = () => {
    this.props.deleteList(this.props.list.id)
    this.props.closeModal()
    this.props.navigation.openDrawer()
  }

  render() {
    return (
      <View style={gStyle.f1}>
        {
          this.props.isModalOpen &&
            <Modal
              closeModal={this.props.closeModal}
              deleteList={this.deleteList}
            />
        }

        {
          this.props.lists.length === 0 &&
            <View style={[gStyle.fcenter, gStyle.f1, style.emptyState]}>
              <SurprisedPaper />
              <Text style={style.emptyStateText}>
                Woah, you don&apos;t have any lists yet.
              </Text>
            </View>
        }

        {
          this.props.lists.length > 0 && !this.props.list &&
            <View style={[gStyle.fcenter, gStyle.f1, style.emptyState]}>
              <HappyPaper />
              <Text style={style.emptyStateText}>
                Hey, go select a list or create a new one.
              </Text>
            </View>
        }

        {
          this.props.list &&
            <List list={this.props.list} />
        }
      </View>
    )
  }
}

const mapStateToProps = ({ lists, globalUi }) => ({
  lists,
  list: lists.find(list => list.id === globalUi.activeListId),
  isModalOpen: globalUi.isModalOpen,
  favouriteListId: globalUi.favouriteListId,
})

const mapDispatchToProps = {
  setActiveList,
  deleteList,
  closeModal: setIsModalOpen.bind(null, false),
}

export default connect(mapStateToProps, mapDispatchToProps)(App)
