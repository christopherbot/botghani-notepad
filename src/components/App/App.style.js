import { StyleSheet } from 'react-native'

export default StyleSheet.create({
  navWrapper: {
    display: 'flex',
    flexDirection: 'row',
    justifyContent: 'flex-start',
    position: 'absolute',
    height: '100%',
    zIndex: 1,
  },
})
