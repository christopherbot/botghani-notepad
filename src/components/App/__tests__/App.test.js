import React from 'react'    
import { App } from '../App'
import renderer from 'react-test-renderer'

describe('App', () => {
  const mockNavigation = {
    openDrawer: jest.fn(),
    closeDrawer: jest.fn(),
  }

  let app, props

  beforeEach(() => {
    props = {
      lists: [{}],
      setActiveList: jest.fn(),
      deleteList: jest.fn(),
      closeModal: jest.fn(),  
      isModalOpen: true,
      navigation: mockNavigation,
    }

    app = renderer.create(<App {...props}/>).toJSON()
  })

  it('should match the render app component', () => {
    expect(app).toMatchSnapshot()
  })
})  
