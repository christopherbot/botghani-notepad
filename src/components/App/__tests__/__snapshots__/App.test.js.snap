// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the render app component 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
        "flexDirection": "column",
      }
    }
  >
    <Modal
      animationType="slide"
      hardwareAccelerated={false}
      onRequestClose={
        [MockFunction] {
          "calls": Array [
            Array [],
          ],
        }
      }
      visible={true}
    >
      <View
        style={
          Array [
            Object {
              "flex": 1,
              "flexDirection": "column",
            },
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#9ff6fd",
              "height": "100%",
              "paddingVertical": 50,
            },
          ]
        }
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "marginBottom": 20,
            }
          }
        >
          Are you sure you want to delete this list?
        </Text>
        <View
          style={
            Object {
              "flexDirection": "row",
              "justifyContent": "space-between",
              "paddingHorizontal": 125,
              "width": "100%",
            }
          }
        >
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              Cancel
            </Text>
          </View>
          <View
            accessibilityComponentType={undefined}
            accessibilityLabel={undefined}
            accessibilityTraits={undefined}
            accessible={true}
            collapsable={undefined}
            hasTVPreferredFocus={undefined}
            hitSlop={undefined}
            isTVSelectable={true}
            nativeID={undefined}
            onLayout={undefined}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "opacity": 1,
              }
            }
            testID={undefined}
            tvParallaxProperties={undefined}
          >
            <Text
              accessible={true}
              allowFontScaling={true}
              ellipsizeMode="tail"
            >
              Yes
            </Text>
          </View>
        </View>
      </View>
    </Modal>
  </View>
</View>
`;
